### PR TITLE
chore: array.sort() intrinsic function for the new ssa

### LINKED ISSUE #1687: Implement `Intrinsic::Sort`
### Problem

The test named `array_len` should compile. Instead SSA gen errors:
```
The application panicked (crashed).
Message:  not yet implemented: expected a black box function
Location: crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs:556
```

### Happy Case

(Opened for tracking - suggestions to be documented later.)

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

Enable the array.sort() feature for the new ssa.

## Problem\*

Resolves #1687 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by jfecher — COMMENTED** [review]
LGTM but I'll leave approval to @kevaundray for reviewing the permutation chages

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs [hunk 1]
```diff
@@ -684,6 +684,26 @@ impl GeneratedAcir {
         });
         self.push_opcode(opcode);
     }
+
+    // Generate gates and control bits witnesses which ensure that out_expr is a permutation of in_expr
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs:688]: What are control bits?

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs [hunk 2]
```diff
@@ -793,6 +795,47 @@ impl AcirContext {
...
+            self.less_than_constrain(outputs_var[i], outputs_var[i + 1], bit_size)?;
+        }
+        // Enforce the outputs to be a permutation of the inputs
+        self.acir_ir.permutation(&inputs_expr, &output_expr);
+
+        Ok(outputs_var)
+    }
+
+    /// Constrain lhs to be less than rhs
+    fn less_than_constrain(
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs:829]: This is pretty confusing, given that we have `less_than_var` -- why can't you use that method?

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs [hunk 3]
```diff
@@ -368,7 +368,7 @@ impl GeneratedAcir {
     /// n.b: we do NOT check here that lhs and rhs are indeed 'bits' size
     /// lhs < rhs <=> a+1<=b
     /// TODO: Consolidate this with bounds_check function.
-    fn bound_constraint_with_offset(
+    pub(crate) fn bound_constraint_with_offset(
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs:371]: Would prefer to keep this private and we wrap it in the higher level functionality we need. 

For example, in this PR, we are using it to do a less_than, in that case, we should expose a less_than function. This is if we exclude the fact that we already have a less_than_var method

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs (hunk 2 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs:829]: Done

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs (hunk 3 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs]: Done

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs (hunk 1 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs]: I added a comment

**Review by kevaundray — APPROVED** [review]

**Pull request merged**